EIPS/eip-7848.md

```diff
@@ -0,0 +1,122 @@
+---
+eip: 7848
+title: On-chain upgrade signaling
+description: Allows participants to indicate readiness for a client upgrade when producing blocks
+author: William Entriken (@fulldecent)
+discussions-to: https://ethereum-magicians.org/t/eip-xxxx-on-chain-upgrade-signaling/22306
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-12-22
+---
+
+## Abstract
+
+This proposal adds a mechanism for clients to signal their willingness for a network upgrade by including a “reference implementation hash” field in each block. A network upgrade activates only if enough blocks specify that they are using the new software.
+
+## Motivation
+
+Currently, upgrades to Ethereum Mainnet are announced on the ethereum.org blog. This proposal changes that process by activating upgrades based on the consent of network participants.
+
+## Specification
+
+Ethereum consensus clients shall identify with a reference implementation having its equivalent behaviors.
+
+Ethereum consensus clients shall include a `referenceImplementationHash` (specified below) field immediately following the `extraData` field in the block header.
+
+Proposed new Ethereum consensus clients shall initially behave as the then-current network does. If, and only if, a successful upgrade (defined below) is activated will the new behavior take effect.
+
+When a network participant is willing to support a network upgrade, they shall immediately begin using the new software. The blocks they create including the new `referenceImplementationHash` will contribute to the network upgrade activation.
+
+### The Reference Implementation Hash
+
+When proposing a network upgrade, the proposer shall point to a published, feature-complete reference implementation including the new software behaviors.
+
+The **reference implementation hash** is the SHA-256 hash of the tarred source code of this reference implementation.
+
+### Verification
+
+Network participants shall study the reference implementation and decide whether they support an upgrade.
+
+Network participants shall study any specific software they will run and ensure it faithfully implements the behaviors consistent with that referennce implementation.
+
+### Upgrade
+
+Upgrade proposals (often referred to as hard fork EIPs) must specify an upgrade window and threshold. These parameters are implemented in the consensus client:
+
+- `VOTING_WINDOW_BEGIN`: the first block (inclusive) to count votes.
+- `VOTING_WINDOW_END`: the last block (inclusive) to count votes.
+- `REFERENCE_IMPLEMENTATION_HASH`: this is defined above.
+- `REQUIRED_APPROVALS`: the minimum number of approvals necessary for the upgrade.
+
+The difference (`VOTING_WINDOW_END` - `VOTING_WINDOW_BEGIN`) must be strictly greater than 14400 (about two days) and should be chosen to provide sufficient time to advertise the upgrade to network participants.
```

Comment:
could you add details on how many fraction of validators will actually get to vote which is the fraction of validators having the block proposal in that particular window.

also add the justification for that particular fraction in the rational section.

Reply:
This proposal intentionally does not specify the fraction of validators.

Somebody using this EIP is responsible to chose the voting window. Their reasons may include a fraction of validators, or other reasons.

---

The note about two days is a strict lower bound. This is included only for a human persective because humans can sleep for up to one day at a time. And two is more than one.

It is actually expected that upgrades using this EIP will be MUCH more than two days, for safety.

Reply:
could you please specify it or add <-- TODO --> sections

Reply:
It is not a to do.

It is not specified and that is an intentional design decision of this EIP not to overspecify.

---

As currently written, there are "`REQUIRED_APPROVALS`: the minimum number of approvals necessary for the upgrade." And also, upgrade occurs if "`REQUIRED_APPROVALS` or more blocks during the voting window had set `referenceImplementationHash` to the value `REFERENCE_IMPLEMENTATION_HASH`".

---

This is specified. Open for discussion. But that discussion should be after passing draft status.

Reply:
the reason I want you to do calculation is because you have specified 2 day window and you need to calculate how many validators will vote and what fraction of that will actually be voting

and since that is going to be very small fraction, I want to see your reasoning for deciding an upgrade based on a very small fraction

Reply:
Here's the complete reasoning:

Previously Ethereum Foundation mandated a change to the network with less than 24 hours notice. 

These changes must be implemented by humans and humans typically sleep on a 24-hour schedule.

What EIP did previously was unacceptable. This EIP codifies that it is unacceptable by mandating a minimum of two days.

No math or discussion of validation or epochs or slots is necessary. A basic understanding of human circadian rhythms on earth is all that is needed here.

---

Additionally, I agree with you that it will be helpful to think about validators and the percentage of the pool and all of that. But that is not part of this specification. It is an implementation detail that will need to be decided each time this specification is used.